### ISSUE TITLE
salt-cloud: Exception raised when processing __virtual__ function

### ISSUE THREAD

**Issue by kshenk1** [saltstack/salt#24579]
Just upgraded to 2015.5.1 - also reproduced on 2015.5.2

Using salt-cloud with aws ec2. Installing from git tag

```
[root@salt-master-01 ~]# uname -a
Linux salt-master-01 3.10.0-123.20.1.el7.x86_64 #1 SMP Thu Jan 29 18:05:33 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
[root@salt-master-01 ~]# cat /etc/redhat-release
CentOS Linux release 7.0.1406 (Core)
```

```
[root@salt-master-01 ~]# salt --versions-report
           Salt: 2015.5.1
         Python: 2.7.5 (default, Jun 17 2014, 18:11:42)
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.5.0
           RAET: Not Installed
            ZMQ: 4.0.5
           Mako: Not Installed
```

And this is what I get after salt's installed
```
Writing /usr/lib/python2.7/site-packages/salt-2015.5.1-py2.7.egg-info
 *  INFO: Running install_centos_git_post()
ln -s '/usr/lib/systemd/system/salt-minion.service' '/etc/systemd/system/multi-user.target.wants/salt-minion.service'
 *  INFO: Running install_centos_check_services()
 *  INFO: Running install_centos_restart_daemons()
 *  INFO: Running daemons_running()
 *  INFO: Salt installed!
Warning: Permanently added '1.2.3.4' (ECDSA) to the list of known hosts.
Connection to 1.2.3.4 closed.
Warning: Permanently added '1.2.3.4' (ECDSA) to the list of known hosts.
Connection to 1.2.3.4 closed.
Warning: Permanently added '1.2.3.4' (ECDSA) to the list of known hosts.
Connection to 1.2.3.4 closed.
Warning: Permanently added '1.2.3.4' (ECDSA) to the list of known hosts.
Connection to 1.2.3.4 closed.
Warning: Permanently added '1.2.3.4' (ECDSA) to the list of known hosts.
Connection to 1.2.3.4 closed.
[INFO    ] Salt installed on webapp-00.env51
[INFO    ] Created Cloud VM 'webapp-00.env51'
[ERROR   ] Exception raised when processing __virtual__ function for selinux. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.selinux.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'selinux', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for groupadd. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.groupadd.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'groupadd', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for freebsdkmod. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.freebsdkmod.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'freebsdkmod', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for smartos_vmadm. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.smartos_vmadm.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'smartos_vmadm', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for systemd. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.systemd.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'systemd', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for layman. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.layman.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'layman', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for freebsdservice. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.freebsdservice.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'freebsdservice', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for openbsdrcctl. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.openbsdrcctl.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'openbsdrcctl', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for linux_sysctl. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.linux_sysctl.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'linux_sysctl', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for rh_service. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.rh_service.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'rh_service', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for launchctl. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.launchctl.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'launchctl', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for service. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.service.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'service', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for rh_ip. Module will not be loaded 'os_family'
[WARNING ] salt.loaded.int.module.rh_ip.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'rh_ip', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for pkgng. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.pkgng.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'pkgng', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for rest_package. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.rest_package.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'rest_package', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for solarispkg. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.solarispkg.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'solarispkg', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for debian_ip. Module will not be loaded 'os_family'
[WARNING ] salt.loaded.int.module.debian_ip.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'debian_ip', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for pacman. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.pacman.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'pacman', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for freebsdjail. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.freebsdjail.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'freebsdjail', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for openbsd_sysctl. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.openbsd_sysctl.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'openbsd_sysctl', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for pw_user. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.pw_user.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'pw_user', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for smf. Module will not be loaded 'os_family'
[WARNING ] salt.loaded.int.module.smf.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'smf', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for makeconf. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.makeconf.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'makeconf', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for pkgin. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.pkgin.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'pkgin', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for solarisips. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.solarisips.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'solarisips', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for pw_group. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.pw_group.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'pw_group', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for smartos_imgadm. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.smartos_imgadm.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'smartos_imgadm', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for freebsd_sysctl. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.freebsd_sysctl.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'freebsd_sysctl', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for openbsdpkg. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.openbsdpkg.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'openbsdpkg', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for xapi. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.xapi.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'xapi', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for apache. Module will not be loaded 'os_family'
[WARNING ] salt.loaded.int.module.apache.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'apache', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for chocolatey. Module will not be loaded 'os_family'
[WARNING ] salt.loaded.int.module.chocolatey.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'chocolatey', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for pkgutil. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.pkgutil.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'pkgutil', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for openbsdservice. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.openbsdservice.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'openbsdservice', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for solaris_group. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.solaris_group.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'solaris_group', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for eselect. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.eselect.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'eselect', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for debian_service. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.debian_service.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'debian_service', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for rest_service. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.rest_service.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'rest_service', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for netbsd_sysctl. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.netbsd_sysctl.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'netbsd_sysctl', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for deb_apache. Module will not be loaded 'os_family'
[WARNING ] salt.loaded.int.module.deb_apache.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'deb_apache', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for useradd. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.useradd.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'useradd', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for gentoo_service. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.gentoo_service.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'gentoo_service', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for dpkg. Module will not be loaded 'os_family'
[WARNING ] salt.loaded.int.module.dpkg.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'dpkg', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for grub_legacy. Module will not be loaded 'os_family'
[WARNING ] salt.loaded.int.module.grub_legacy.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'grub_legacy', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for eix. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.eix.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'eix', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for freebsdpkg. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.freebsdpkg.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'freebsdpkg', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for zfs. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.zfs.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'zfs', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for debconfmod. Module will not be loaded 'os_family'
[WARNING ] salt.loaded.int.module.debconfmod.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'debconfmod', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for bridge. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.bridge.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'bridge', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for netbsdservice. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.netbsdservice.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'netbsdservice', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for darwin_sysctl. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.darwin_sysctl.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'darwin_sysctl', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for mdadm. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.mdadm.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'mdadm', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for poudriere. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.poudriere.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'poudriere', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for kmod. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.kmod.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'kmod', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for solaris_user. Module will not be loaded 'kernel'
[WARNING ] salt.loaded.int.module.solaris_user.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'solaris_user', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for softwareupdate. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.softwareupdate.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'softwareupdate', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for gentoolkitmod. Module will not be loaded 'os'
[WARNING ] salt.loaded.int.module.gentoolkitmod.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'gentoolkitmod', please fix this.
[ERROR   ] Exception raised when processing __virtual__ function for logadm. Module will not be loaded 'os_family'
[WARNING ] salt.loaded.int.module.logadm.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'logadm', please fix this.
[ERROR   ] There was a profile error: unhashable type: 'dict'
```

And here's my cloud profile that worked on 2014.1.x, and 2014.7.x
```
web:
  provider: ec2
  image: ami-1234
  size: c3.xlarge
  ssh_username: root
  script_args: -H http://proxy.svcs:3128 -G -Z git v2015.5.1
  network_interfaces:
    - DeviceIndex: 0
      PrivateIpAddresses:
        - Primary: True
      AssociatePublicIpAddress: False
      SubnetId: subnet-1234
      SecurityGroupId:
        - sg-1234
        - sg-2345
        - sg-3456
  minion:
    master: salt-master.admin
  ebs_optimized: False
  grains:
    stuff: things
```

And the provider.....
```
ec2:
  provider: ec2
  id: theid
  key: 'the key'
  keyname: my-key
  location: us-west-1
  rename_on_destroy: True
  ssh_interface: private_ips
  ssh_username: root
  private_key: /path/to/file.pem
```

Please tell me I'm doing something wrong, and there's just something small I need to adjust. I was really looking forward to this......

**Comment by kshenk1** [comment]
I found the issue, and found a way to get past this. For anyone else that might be in a similar boat....

We have a .conf file in `/etc/salt/cloud.conf.d` which contains `sync_after_install: all`
We also have the reactor setup to handle `'salt/cloud/*/created'` events. The SLS that is called from that event needs to restart the minion, which was apparently directly interfering with syncing any and all custom grains, modules...etc.

*salt.loaded.int.module.`something`.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name.*

That seemed odd - especially if you look at any one of the modules that blew chunks. So I'm thinking that maybe python was running the code while the salt-minion was in the middle of the restart.

Either way, I just removed the `sync_after_install: all` from `/etc/salt/cloud.conf.d/file.conf`, and placed the following in the SLS triggered by the `salt/cloud/*/created` event

```
sync_all:
  local.saltutil.sync_all:
    - tgt: {{ data['name'] }}
```
